src/enzyme_ad/jax/Passes/EnzymeHLOOpt.cpp

```diff
@@ -15414,6 +15414,133 @@ struct ConcatConcatAxisSwap final
+    }
+
+    Value baseOperand = triggerSliceOp.getOperand();
+
+    // --- Get Target Rotate Parameters ---
+    int targetAmount = triggerRotateOp.getAmount();
+    int targetRotateDim = triggerRotateOp.getDimension();
+    Location loc = triggerRotateOp.getLoc();
+
+    // --- Check Validity of Trigger Slice ---
+    auto baseOperandType = dyn_cast<RankedTensorType>(baseOperand.getType());
+    if (!baseOperandType || !baseOperandType.hasStaticShape()) {
+      return rewriter.notifyMatchFailure(triggerRotateOp,
+                                         "Base operand requires static shape");
+    }
+    if (triggerSliceOp.getStartIndices()[targetRotateDim] != 0 ||
+        triggerSliceOp.getLimitIndices()[targetRotateDim] !=
+            baseOperandType.getShape()[targetRotateDim] ||
+        triggerSliceOp.getStrides()[targetRotateDim] != 1) {
+      return rewriter.notifyMatchFailure(
+          triggerRotateOp,
+          "Trigger SliceOp modifies the dimension being extended");
+    }
+
+    llvm::SmallVector<CandidateInfo> candidates;
+    candidates.push_back({triggerRotateOp, triggerSliceOp});
+
+    for (auto const &userOp : baseOperand.getUsers()) {
+      // Skip the slice that defines the trigger operand
+      if (userOp == triggerSliceOp.getOperation())
+        continue;
+
+      // Case 1: Direct Rotate
+      if (auto directRotate = dyn_cast<enzymexla::RotateOp>(userOp)) {
+        if (directRotate.getDimension() == targetRotateDim &&
+            directRotate.getAmount() == targetAmount) {
+          candidates.push_back({directRotate, nullptr});
+        }
+      }
+      // Case 2: Rotate of Slice
+      else if (auto sliceUser = dyn_cast<stablehlo::SliceOp>(userOp)) {
+        if (!sliceUser->hasOneUse())
+          continue;
+        auto rotateOfSlice =
+            dyn_cast<enzymexla::RotateOp>(*sliceUser->user_begin());
+        if (!rotateOfSlice)
+          continue;
+
+        if (rotateOfSlice.getDimension() == targetRotateDim &&
+            rotateOfSlice.getAmount() == targetAmount) {
+          // Check validity: sliceUser must not modify targetRotateDim
+          if (sliceUser.getStartIndices()[targetRotateDim] == 0 &&
+              sliceUser.getLimitIndices()[targetRotateDim] ==
+                  baseOperandType.getShape()[targetRotateDim] &&
+              sliceUser.getStrides()[targetRotateDim] == 1) {
+            candidates.push_back({rotateOfSlice, sliceUser});
+          }
+        }
+      }
+    }
+
+    if (candidates.size() <= 1) {
+      return rewriter.notifyMatchFailure(
+          triggerRotateOp,
+          "Rewrite condition not met (only found the trigger candidate)");
+    }
+
+    SmallVector<int64_t> newBaseExtendShape =
+        llvm::to_vector(baseOperandType.getShape());
+
+    auto newBaseRotateType = baseOperandType;
+
+    auto newBaseRotateOp = rewriter.create<enzymexla::RotateOp>(
+        loc, newBaseRotateType, baseOperand, targetAmount, targetRotateDim);
```

Comment:
same comment about correctness from https://github.com/EnzymeAD/Enzyme-JAX/commit/53652ee51583d80d5e016836e86e1ccc649a34ba

Reply:
I added a test, and one for slicewrap as well